### PR TITLE
Remove value syncing in Link Control

### LINKED ISSUE #51256: Remove value sync hook in Link Control
Let's look to remove the value sync that was only needed to fix an obscure bug that has other guards.

-----

Do you know if this sync hook is still needed? Now that value is an object; the synchronization can cause the content loss.

_Originally posted by @Mamaduka in https://github.com/WordPress/gutenberg/pull/50668#discussion_r1210710970_

### PR DESCRIPTION
**Pull request by getdave**
## What?

Removes the (potentially) unwanted value syncing in Link Control component.

Closes https://github.com/WordPress/gutenberg/issues/51256

## Why?

It might be unnecessary as per https://github.com/WordPress/gutenberg/issues/51256

## How?

Remove the effectful code.

## Testing Instructions

Run the tests with

```
npm run test:unit packages/block-editor/src/components/link-control/
```

### CODE REVIEW

**Comment by getdave** [comment]
@Mamaduka if you look at the scenario in:

```
npm run test:unit packages/block-editor/src/components/link-control/ -- --t "should reset state upon controlled value change"
```

...it should outline why this code exists. Let's see if we can work out a way to remove it.

For example in rich text we use a `key` to force remount:

https://github.com/WordPress/gutenberg/blob/8db07c17a1329a09b7a92ac8a7817e14d61885f4/packages/format-library/src/link/inline.js#L261-L264

**Review by Mamaduka — COMMENTED** [review]
The `key` is one of the recommended ways to reset the component state in React. So I don't see anything wrong with using it.

Maybe we just need to better document this method for `LinkControl`.

cc @kevin940726

**Comment by getdave** [comment]
@Mamaduka One approach we could consider here is exporting a wrapped version of `LinkControl` which has a `key` generated for you based on the `value` prop.

```js
const LinkControlWithKey = ( props ) => {
	return (
		<LinkControl { ...props } key={ useLinkInstanceKey( props?.value ) } />
	);
};

export default LinkControlWithKey;
```

I tried this and it does seem to work. The hook comes from

https://github.com/WordPress/gutenberg/blob/7eddee3be86dbeefc06990d817768e15de42d9fc/packages/format-library/src/link/use-link-instance-key.js#L19

I'm not 100% sure about the approach so I'd appreciate your 👀 (and also @kevin940726 if he is around?).

**Comment by Mamaduka** [comment]
My suggestion would suggest not to use an implicit approach. Let the consumer components handle state resetting via `key`. The less "magic" is there inside the components, the easier it's to debug them :)

We can document the case in the readme file and link the React docs. What do you think?

**Comment by getdave** [comment]
> My suggestion would suggest not to use an implicit approach. Let the consumer components handle state resetting via `key`. The less "magic" is there inside the components, the easier it's to debug them :)
> 
> We can document the case in the readme file and link the React docs. What do you think?

That's probably best. I just realised there's a problem with the automatic approach anyway. Ok so we remove it and update docs. I'll write the docs tomorrow UTC.

**Comment by getdave** [comment]
@Mamaduka I updated the docs. Let's refine them and see if it's good enough to merge now I removed the test coverage.

**Comment by Mamaduka** [comment]
Thank you, @getdave! The update looks good (for me, personally) but also feels too technical. People who'll understand all the details there probably know about resetting the state via `key`.

Maybe we should just document how to solve this particular case and skip why it might happen.

P.S. I'm terrible at writing docs. Going to ping @ndiego and @ryanwelcher if they have any suggestions.

**Review by kevin940726 — DISMISSED** [review]
I'm a bit confused as it's been a while since I looked at `<LinkControl>` so excuse me if I'm horribly wrong 😅 . Don't we already have `key` defined in https://github.com/WordPress/gutenberg/blob/fa4f5a5e4e5a1c1e1309f59ff9a99526f92a0a6a/packages/block-editor/src/components/link-control/index.js#L372? Is there a case where that's not enough? Should we just update the key to take the full value (rather than only the url` into account?

Either way I think this is still a better change. Nice job on updating the doc too! Perhaps we can also link to the official React doc about this for more info?

**Comment by getdave** [comment]
@kevin940726 Thanks for your review and feedback. Much appreciated.

That `key` is just for the _preview_ of the Link. The component itself doesn't have a `key` set internally.

However certain implementations such as the rich text implementation of Link UI does:

https://github.com/WordPress/gutenberg/blob/fa4f5a5e4e5a1c1e1309f59ff9a99526f92a0a6a/packages/format-library/src/link/inline.js#L262-L263

This PR is suggesting making such an implementation the "best practice" and not trying to be clever about syncing values.

**Comment by getdave** [comment]
Ok just waiting on reviews from documentation experts. If all is 👍 please feel free to merge.

**Comment by getdave** [comment]
Potential bug highlighted by the failing e2e tests here.

**Comment by ndiego** [comment]
> The update looks good (for me, personally) but also feels too technical. People who'll understand all the details there probably know about resetting the state via key.

Personally, I like the extra details. While I can comfortably build "stuff" with React, my knowledge of React fundamentals is not comprehensive, which I imagine is the case for many other developers. The added insight was helpful. 

Unless @ryanwelcher feels differently, I would move forward.

**Comment by scruffian** [comment]
The test failure is genuine. It happens when we update a link value in the button component and then that needs to get synced back to link control. I'm not sure if that means we need this code after all, or if there's another way to achieve that.

**Comment by getdave** [comment]
> The test failure is genuine. It happens when we update a link value in the button component and then that needs to get synced back to link control. I'm not sure if that means we need this code after all, or if there's another way to achieve that.

Ok let's look into that...thanks for confirming the issue @scruffian 🙇

**Comment by getdave** [comment]
I've taken a look at the test failure and it does indeed seem legit.

```
npm run test:e2e:playwright -- -g "appends http protocol to links added which are missing a protocol"
```

The issue is that as @scruffian described. The buttons block uses `prependHttp` to modify the `url` of the new value object passed to the `onChange` callback by `LinkControl`:

https://github.com/WordPress/gutenberg/blob/e57bf0dd418ca70dda339c2bc2cca41a90b2182c/packages/block-library/src/button/edit.js#L239-L248

This causes the value passed as `value` to change and become out of sync with the internal value. This is why we have the effect to sync the internal value back up with newly provided value.

The component is a controlled component so you should be able to modify the provided `value` and see that reflected in the component. However, we are also maintaining some internal state (which is kinda of a hybrid approach) in order that the component can control if/when it can be submitted. Traditionally the consumer would handle that kind of logic but it would place a burden on them.

Another reason for the internal state is to handle things like checkboxes for `Open in new tab` which shouldn't modify the `value` unless the _user_ specifically chooses to `Save` the changes in the UI. Previously we had a lot of bugs whereby users would toggle `Opens in new tab` and find their changes applied unexpectedly.

The only way I can see this working is if we remove the internal state and make the component fully controlled. This would implement IoC but have the consequence of placing the burden of predictable behaviour onto the consumer. For example, they would need to implement all the logic to determine if/when a value is valid and can be submitted.

One way around this might be to expose some standardised methods on the component (`LinkControl.isSubmittable` & `LinkControl.isValid` ) which _consumers_ could call in their `onChange` callback to determine whether the controlled `value` should be updated. This would be a breaking change however, because the component would start to behave differently from this point forward requiring all 3rd party consumers to update their implementations.

The benefits of such a change would be we remove this hybrid modal and get a predictable controlled component. The downsides are breaking backwards compatibility and also placing additional burden on the consumer.  

Does anyone have any better suggestions? Which route should we take?

**Comment by Mamaduka** [comment]
Thank you, @getdave!

Unfortunately, I don't have any bright ideas at the moment. The explorations here should show us the pain points of the component better, which would be helpful in the future.

Let's keep this PR or issue open for now, and I'll try to experiment with "value syncing" removal more when I get a chance.

P.S. I think prepending the HTTP logic could be absorbed by the `LinkControl` component. I thought it was already doing that, but it looks like [Link format also handles](https://github.com/WordPress/gutenberg/blob/79dbfc57a0f3918a299f5856f69561f695ff8896/packages/format-library/src/link/inline.js#L112) it separately like the Button.

**Comment by getdave** [comment]
> I think prepending the HTTP logic could be absorbed by the `LinkControl` component. I thought it was already doing that, but it looks like [Link format also handles](https://github.com/WordPress/gutenberg/blob/79dbfc57a0f3918a299f5856f69561f695ff8896/packages/format-library/src/link/inline.js#L112) it separately like the Button.

Yes it seems like this is key functionality which could be absorbed. As long as you and I are careful that it's not seen as a solution to allowing this PR to merge though 😓

**Review by ajlende — DISMISSED** [review]
Leaving a blocking review so it isn't merged before the HTTP logic is absorbed by the `LinkControl` component.

**Comment by getdave** [comment]
@Mamaduka I've been thinking about this. I think one option could be to get rid of this internal value code and instead make this a true controlled input.

In practice this would mean the following API:

- `onChange` - called on any modification to the `value` via the control. This includes typing in the inputs or toggling controls. Does _not_ signal that the user has committed their changes.
- `onSubmit` - called when the user actively commits their changes. 

So consumers would wire things up as follows:

```jsx
const value = {
    // existing link value from persistence layer somewhere.
    // for example block attributes
};

const [newLink, setNewLink] = useState(value);

function persistLink(val) {
    // function which updates the persistence layer with
    // the new link.
}


return (
    <LinkControl
        value={newLink}
        onChange={setNewLink}
        onSubmit={persistLink}
    />
)
```

The benefit is that whilst the consumer accepts responsibility for wiring up the temporary state of the link value (via `onChange`) they can be sure that the component will only call `onSubmit` when the user wants to commit changes. That means they don't have to worry about anything to do with the UI/UX in the data handling logic. All they need do is fulfill the contract of `onChange` and `onSubmit`.

The problem I see with the approach is backwards compatibility. Up until now `onChange` has implied the value is being committed as we handle the temporary changes internally within the component (thus the value syncing).

We have two options:

- We change the component API and accept potential for breaking change - this could be said to be legitimate because the component is `__experimental`. However in practice it has been so widely used that this would probably not be a good approach.
- We make `onChange` effectively `onSubmit` and add a new property `onTempChange` to handle the temporary values. This would need to be well tested!

**Questions**:

- what do we gain from removing internal value syncing and making the component controlled?
- is backwards compatibility possible?
- is this approach worth the effort?

**Comment by Mamaduka** [comment]
Thank you, @getdave!

As you mentioned in your previous comment, that burdens consumers more. While I don't think it's an issue if we have used this method from the start, now it might be perceived as a significant breaking change. We have to remember that consumers have to support multiple versions of WP.

Let's make this a low priority for now; the gain probably isn't worth it, and the annoying issues could be solved by memoizing the `value` passed to a component (#53507).

What do you think?

**Comment by getdave** [comment]
> ...burdens consumers more. While I don't think it's an issue if we have used this method from the start, now it might be perceived as a significant breaking change. We have to remember that consumers have to support multiple versions of WP.

Exactly. This is the major problem. I think the advantages aren't worth it, especially with the backwards compatibility concerns I outlined before.

> Let's make this a low priority for now; the gain probably isn't worth it, and the annoying issues could be solved by memoizing the `value` passed to a component (#53507).

👍 I think we should update the documentation for the component to recommend this approach.

**Comment by kevin940726** [comment]
Sorry for coming back to this so late! It seems like changing the API dramatically is not worth it at this stage for backward-compatibility reasons. Using `key` isn't always the best way either as it'll sometimes unmount the component and cause the link to lose focus, and it's a rabbit hole to fix it all at this stage. It feels to me that we want to keep the internal syncing, but maybe we can find a quick way to avoid syncing with `useEffect`, which causes double-render.

I propose deriving states by [storing states from the previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders) instead:

```diff
- useEffect( () => {
- 	setInternalValue( ( prevValue ) => {
- 		if ( value && value !== prevValue ) {
- 			return value;
- 		}
- 
- 		return prevValue;
- 	} );
- }, [ value ] );
+ const [ previousValue, setPreviousValue ] = useState( value );
+ 
+ if ( value !== previousValue ) {
+ 	setPreviousValue( value );
+ 	if ( value !== internalValue ) {
+ 		setInternalValue( value );
+ 	}
+ }
```

It seems to work in my testing, WDYT?

**Comment by getdave** [comment]
> It seems to work in my testing, WDYT?

I've tested this and it appears to work as described 🥳 

The unit test below will pass as well:

```
npm run test:unit packages/block-editor/src/components/link-control/ -- --t="should reset state upon controlled value change"
```

File: packages/block-editor/src/components/link-control/README.md
Comment (kevin940726) [inline packages/block-editor/src/components/link-control/README.md:15]: This could be slightly misleading. React only does that when the component node is at the same place in the tree. This happens in gutenberg because we use `SlotFill` to portal the component to different places, which in the meantime tricks the React reconciler into thinking that they are the same node.

**Review by kevin940726 — APPROVED** [review]
Thank you! Looks good! The e2e failure doesn't seem related, that test has been quite flaky for a while.

File: packages/block-editor/src/components/link-control/README.md
Comment (getdave) [reply packages/block-editor/src/components/link-control/README.md:15]: Questions:

- How would you recommend we improve this section of the README? 
- Shall we just remove this stuff about `key`? Or do you think it remains relevant (I think it is)?

**Comment by Mamaduka** [comment]
I'm not sure about the new solution. It doesn't do much if the component doesn't receive a stable value reference; when the value changes, both states will be updated. Basically, we're just replacing effect with state update during render.

Yes, it fixes double re-rendering, but not the original issue.

**Comment by getdave** [comment]
Pinging @Mamaduka for a final 👀 before we merge this one. 

I suggest that if merged we  do _not_ include this in WP 6.4. That will allow sufficient time for any unintended consequences of the changes to filter through outside of a major release BETA period.

It can then safely land in 6.5.

File: packages/block-editor/src/components/link-control/README.md
Comment (kevin940726) [reply packages/block-editor/src/components/link-control/README.md:15]: I don't have a preference TBH. I think it's okay to keep this, but maybe we can also make it less misleading. We could also add a link to the official React doc about reconciling for reference.

**Comment by kevin940726** [comment]
> I'm not sure about the new solution. It doesn't do much if the component doesn't receive a stable value reference; when the value changes, both states will be updated. Basically, we're just replacing effect with state update during render.

Yep! It's simply a replacement for the `useEffect`. I don't think it's _ideal_, but still better than the `useEffect` IMHO.

> Yes, it fixes double re-rendering, but not the original issue.

Is the original issue about `value` being an object rather than a stable primitive value? I'm not sure if we can solve it without introducing breaking change, if I'm not mistaken. I think it's okay to be a follow-up if needed. This PR is already a great improvement, and maybe a new thread exploring the potential ideas could freshen up our minds 😉! That is, we can merge this without closing https://github.com/WordPress/gutenberg/issues/51256. And perhaps update the original issue description to clarify things too?

**Comment by Mamaduka** [comment]
I guess we can also do a deep equality checks and only sync state then. WDYT?

But I'm also okay with current solution.

Example:

```js
if ( ! fastDeepEqual( value, internalValue ) ) {
    setInternalValue( value );
}
```

**Comment by kevin940726** [comment]
> I guess we can also do a deep equality checks and only sync state then. WDYT?

Sounds good to me too! 💯

**Comment by Mamaduka** [comment]
Okay, I've adjusted a snippet a little bit. We want to keep the previous value snapshot in a separate state and compare it against it as the internal state changes based on user input.


```js
if ( ! fastDeepEqual( value, previousValue ) ) {
    setPreviousValue( value );
    setInternalValue( value );
}
```

**Comment by getdave** [comment]
> Okay, I've adjusted a snippet a little bit. We want to keep the previous value snapshot in a separate state and compare it against it as the internal state changes based on user input.
> 
> ```js
> if ( ! fastDeepEqual( value, previousValue ) ) {
>     setPreviousValue( value );
>     setInternalValue( value );
> }
> ```

I've applied this change and tests are ✅ 

Specifically this test

```
npm run test:unit packages/block-editor/src/components/link-control -- -t="should reset state upon controlled value change"
```

**Comment by Mamaduka** [comment]
Thanks, @getdave! Let's see if all tests are passing, including e2e, and then I think we're good to merge.

**Pull request merged**
